quinn/tests/many_connections.rs

```diff
@@ -0,0 +1,203 @@
+    let (conn_done_tx, conn_done_rx) = mpsc::unbounded();
+
+    for _ in 0..ctx.borrow_mut().expected_messages {
+        let data = random_data_with_hash(1024 * 1024);
+        let ctx = ctx.clone();
+        let conn_done_tx = conn_done_tx.clone();
+
+        let task = unwrap!(endpoint.connect_with(&client_cfg, &listener_addr, "Test"))
+            .map_err(|e| panic!("Connection failed: {}", e))
+            .and_then(move |(conn_driver, conn, _)| {
+                current_thread::spawn(conn_driver.then(move |_| {
+                    // connection was closed, notify about it
+                    let _ = unwrap!(conn_done_tx.unbounded_send(()));
+                    Ok(())
+                }));
+                write_to_peer(&conn, data);
+                ctx.borrow_mut().connections.push(conn);
+                Ok(())
+            });
+        runtime.spawn(task);
+    }
+
+    let wait_for_all_connections = conn_done_rx
+        .take(ctx.borrow().expected_messages as u64)
+        .for_each(|_| Ok(()));
+    unwrap!(runtime.block_on(wait_for_all_connections));
+
+    let ctx = ctx.borrow();
+    assert_eq!(ctx.received_messages, ctx.expected_messages);
+}
+
+fn read_from_peer(
+    stream: quinn::NewStream,
+    ctx: Rc<RefCell<TestContext>>,
+    conn: quinn::Connection,
+) {
+    let stream = match stream {
+        quinn::NewStream::Bi(_bi) => panic!("Unexpected bidirectional stream here"),
+        quinn::NewStream::Uni(uni) => uni,
+    };
+
+    let task = quinn::read_to_end(stream, 1024 * 1024 * 5)
+        .map_err(|e| panic!("read_to_end() failed: {}", e))
+        .and_then(move |(_stream, data)| {
+            assert!(hash_correct(&data));
+            ctx.borrow_mut().received_messages += 1;
+            conn.close(0, &[]);
+            Ok(())
+        })
+        .then(|_| Ok(()));
+    current_thread::spawn(task);
+}
+
+fn write_to_peer(conn: &quinn::Connection, data: Vec<u8>) {
+    let task = conn
+        .open_uni()
+        .map_err(|e| panic!("Failed to open unidirection stream: {}", e))
+        .and_then(move |o_stream| {
+            tokio::io::write_all(o_stream, data).map_err(|e| panic!("write_all() failed: {}", e))
+        })
+        .and_then(move |(o_stream, _)| {
+            tokio::io::shutdown(o_stream).map_err(|_| ()) // ignore errors, remote peer might already have hung up
+        })
+        .map(|_| ());
+    current_thread::spawn(task);
+}
+
+/// Builds client configuration. Trusts given node certificate.
+fn configure_connector(node_cert: &[u8]) -> quinn::ClientConfig {
+    let mut peer_cfg_builder = quinn::ClientConfigBuilder::new();
+    let their_cert = unwrap!(quinn::Certificate::from_der(&node_cert));
+    unwrap!(peer_cfg_builder.add_certificate_authority(their_cert));
+    let mut peer_cfg = peer_cfg_builder.build();
+    let transport_config = unwrap!(Arc::get_mut(&mut peer_cfg.transport));
+    transport_config.idle_timeout = 0;
+    transport_config.keep_alive_interval = 10_000;
```

Comment:
Why disable the idle timeout? Leaving it at something reasonable might be useful for converting the test timeout into a failure. Additionally, keep-alive isn't useful if the idle timeout is disabled.